In [15]:
import pandas as pd
import statsmodels.api as sm
import numpy  as np
from sklearn.preprocessing import StandardScaler

In [16]:
df = pd.read_csv("nb1-insurance.csv")

In [17]:
# Variable objetivo
Y = df["expenses"]
#Variables predictivas
X = df[["age","sex","bmi","children","smoker"]]

In [18]:
columns_names = df.columns.values.tolist()

In [19]:
dummy_sex = pd.get_dummies(df["sex"], prefix="sex").iloc[:,1:]
dummy_smoker = pd.get_dummies(df["smoker"], prefix="smoker").iloc[:,1:]

In [20]:
columns_names = df.columns.values.tolist()
df_new = df[columns_names].join(dummy_sex)
columns_names = df_new.columns.values.tolist()
df_new = df_new[columns_names].join(dummy_smoker)

In [21]:
feature_cols = ["age","bmi","children","sex_male","smoker_yes"]
# Variable objetivo
Y = df_new["expenses"]
#Variables predictivas
X = df_new[feature_cols]

In [22]:
est = sm.OLS(Y, X).fit()

In [23]:
def predict_expenses(age, bmi, children,sex_male,smoker_yes):
    return 198.399764*age  +  37.578382*bmi + 250.841441*children - 707.668406*sex_male + 23386.227638*smoker_yes

In [24]:
df_new["prediction"] = predict_expenses(df_new["age"],df_new["bmi"],df_new["children"],df_new["sex_male"],df_new["smoker_yes"])

In [25]:
df_new.head()

age     sex   bmi  children smoker     region  expenses  sex_male  \
0   19  female  27.9         0    yes  southwest  16884.92         0   
1   18    male  33.8         1     no  southeast   1725.55         1   
2   28    male  33.0         3     no  southeast   4449.46         1   
3   33    male  22.7         0     no  northwest  21984.47         1   
4   32    male  28.9         0     no  northwest   3866.86         1   

   smoker_yes    prediction  
0           1  28204.260012  
1           0   4384.518099  
2           0   6840.135915  
3           0   6692.553077  
4           0   6727.139282

In [41]:
df_new["SSD"] = (df_new["expenses"] - df_new["prediction"])**2

In [42]:
df_new.head()

age     sex   bmi  children smoker     region  expenses  sex_male  \
0   19  female  27.9         0    yes  southwest  16884.92         0   
1   18    male  33.8         1     no  southeast   1725.55         1   
2   28    male  33.0         3     no  southeast   4449.46         1   
3   33    male  22.7         0     no  northwest  21984.47         1   
4   32    male  28.9         0     no  northwest   3866.86         1   

   smoker_yes    prediction           SSD  
0           1  28204.260012  1.281275e+08  
1           0   4384.518099  7.070111e+06  
2           0   6840.135915  5.715331e+06  
3           0   6692.553077  2.338427e+08  
4           0   6727.139282  8.181198e+06

In [43]:
SSD = sum(df_new["SSD"])
SSD

54986740976.63504

In [44]:
RMSE = np.sqrt(SSD/ len(df_new))
RMSE

6410.633273405596

- ¿Cuál es el RMSE de los modelos de regresión realizados en el notebook 1?
    * RMSE = 6410.633273405596
- ¿Cuáles son las principales limites de un modelo de regresión lineal?
    * Si la variable de salida, es decir, la variable que queremos predecir es de valor discreta, la regresión lineal no nos sirve
    * la prediccion puede que no sea tan precisa
- ¿Cuál es la diferencia entre las relaciones de "correlaciones" y "causalidad"? ¿Por qué es importante tener conciencia de la diferencia?
    * Una correlación es una medida o grado de relación entre dos variables
    * Una relación causal entre dos eventos existe si la ocurrencia del primero causa el otro.
    * Una correlación entre dos variables no implica causalidad. Por otro lado, si hay una relación causal entre dos variables, estas deben estar correlacionadas.
    * Es importante entender la diferencia ya que dos variables esten relacionadas no implica que la segunda variable sea causa de la primera, puede que la segunda variable sea producto de otra que no se este analizando, ejemplo: 
        * Un estudio muestra que hay una correlación negativa entre la ansiedad de un estudiante antes de una prueba y el puntaje del estudiante en una prueba. Pero no podemos decir que la ansiedad causa un puntaje más bajo en la prueba; podría haber otras razones, el estudiante quizá no estudió bien, entonces aquí la correlación no implica causalidad.
        * Hay una correlación positiva entre el número de horas que Usted pasa estudiando y la calificación que obtiene. Aquí, si hay causalidad; si Usted pasa más tiempo estudiando, resultará en una calificación más alta.